src/include/common/execution_thread_pool.h

```diff
@@ -0,0 +1,270 @@
+  };
+
+  common::ManagedPointer<DedicatedThreadRegistry> thread_registry_;
+  // Number of NUMA regions
+  // The worker threads
+  common::SharedLatch array_latch_;
+#ifdef __APPLE__
+  int16_t num_regions_ = 1;
+#else
+  int16_t num_regions_ = numa_available() < 0 || static_cast<int16_t>(numa_max_node()) <= 0
+                             ? 1
+                             : static_cast<int16_t>(numa_max_node()) + 1;
+#endif
+  std::vector<std::vector<TerrierThread *>> workers_;
+  std::vector<ExecutionTaskQueue> task_queue_;
+
+  std::atomic<uint32_t> busy_workers_, total_workers_;
+  std::atomic_bool shutting_down_ = false;
+
+  std::mutex task_lock_;
+  std::condition_variable task_cv_;
+
+  void AddThread(DedicatedThreadTask *t) override {
+    auto *thread = static_cast<TerrierThread *>(t);
+    common::SharedLatch::ScopedExclusiveLatch l(&array_latch_);
+    TERRIER_ASSERT(
+        0 <= static_cast<int16_t>(thread->numa_region_) && static_cast<int16_t>(thread->numa_region_) <= num_regions_,
+        "numa region should be in range");
+    auto *vector = &workers_[static_cast<int16_t>(thread->numa_region_)];
+    vector->emplace_back(thread);
+  }
+  void RemoveThread(DedicatedThreadTask *t) override {
+    auto *thread = static_cast<TerrierThread *>(t);
+    common::SharedLatch::ScopedExclusiveLatch l(&array_latch_);
+    TERRIER_ASSERT(
+        0 <= static_cast<int16_t>(thread->numa_region_) && static_cast<int16_t>(thread->numa_region_) <= num_regions_,
+        "numa region should be in range");
+    std::vector<TerrierThread *> *vector = &workers_[static_cast<int16_t>(thread->numa_region_)];
+    auto it = vector->begin();
+    for (; it != vector->end() && *it != thread; ++it) {
+    }
+    if (it != vector->end()) {
+      vector->erase(it);
+    }
+  }
+
+  bool OnThreadRemoval(common::ManagedPointer<DedicatedThreadTask> dedicated_task) override {
+    // we dont want to deplete a numa region while other numa regions have multiple threads to prevent starvation
+    // on the queue for this region
+    //    auto *thread = static_cast<TerrierThread *>(dedicated_task.operator->());
+    //    if (shutting_down_) return true;
+    //
+    //    common::SharedLatch::ScopedSharedLatch l(&array_latch_);
+    //    TERRIER_ASSERT(0 <= static_cast<int16_t>(thread->numa_region_) && static_cast<int16_t>(thread->numa_region_)
+    //    <= num_regions_,"numa region should be in range"); auto *vector =
+    //    &workers_[static_cast<int16_t>(thread->numa_region_)]; TERRIER_ASSERT(!vector->empty(), "if this thread is
+    //    potentially being closed it must be tracked");
+    //    // if there is another thread running in this numa region
+    //    if (vector->size() > 1) return true;
+    //    for (int16_t i = 0; i < num_regions_; i++) {
+    //      vector = &workers_[i];
+    //      // if another numa region has extra threads
+    //      if (vector->size() > 1) return false;
+    //    }
+    // no numa region has multiple threads
```

Comment:
It this the reason why the function is commented out?

Reply:
Yeah we were trying to be smart when we didn't need to